include/pybind11/pybind11.h

```diff
@@ -394,19 +395,19 @@ class cpp_function : public function {
                 if (result.ptr() != PYBIND11_TRY_NEXT_OVERLOAD)
                     break;
             }
-        } catch (const error_already_set &)      {                                                 return nullptr;
-        } catch (const index_error &e)           { PyErr_SetString(PyExc_IndexError,    e.what()); return nullptr;
-        } catch (const value_error &e)           { PyErr_SetString(PyExc_ValueError,    e.what()); return nullptr;
-        } catch (const stop_iteration &e)        { PyErr_SetString(PyExc_StopIteration, e.what()); return nullptr;
-        } catch (const std::bad_alloc &e)        { PyErr_SetString(PyExc_MemoryError,   e.what()); return nullptr;
-        } catch (const std::domain_error &e)     { PyErr_SetString(PyExc_ValueError,    e.what()); return nullptr;
-        } catch (const std::invalid_argument &e) { PyErr_SetString(PyExc_ValueError,    e.what()); return nullptr;
-        } catch (const std::length_error &e)     { PyErr_SetString(PyExc_ValueError,    e.what()); return nullptr;
-        } catch (const std::out_of_range &e)     { PyErr_SetString(PyExc_IndexError,    e.what()); return nullptr;
-        } catch (const std::range_error &e)      { PyErr_SetString(PyExc_ValueError,    e.what()); return nullptr;
-        } catch (const std::exception &e)        { PyErr_SetString(PyExc_RuntimeError,  e.what()); return nullptr;
+        } catch (const error_already_set &) {
+            return nullptr;
         } catch (...) {
-            PyErr_SetString(PyExc_RuntimeError, "Caught an unknown exception!");
+            auto &registered_exception_translators = pybind11::detail::get_internals().registered_exception_translators;
+            for (auto& translator : registered_exception_translators) {
+                try {
+                    translator(std::current_exception());
+                } catch (...) {
+                    continue;
+                }
```

Comment:
Oh, I see.  You need to call `std::rethrow_exception` within the translators anyway so you can use a `try/catch` block to find out what it is.  I agree that makes my proposal unworkable, at least at this level.

But I am still a bit worried about the performance implications for the likely scenario in LSST, where every package that adds an exception would have to add a custom handler to the list, and that results in a lot of re-throwing of exceptions.  I suppose we could have a single custom handler for LSST, though, and have another way to register lightweight handlers there (since all of our exceptions inherit from a common base class and have virtual destructors, we could traverse the inheritance graph with dynamic casts, for instance).

As a side note: I'm not 100% sure what `std::current_exception` will point to here if the translator throws a different exception than the one it received (is it the last exception thrown, or the one in the currently active `catch` block`).  I suspect it must be the latter, but it might be good to clarify that by assigning the result of `std::current_exception` to a variable with the scope you want.